### PR TITLE
feat: Add party GUI

### PR DESCRIPTION
**Pull request by ryanzhoudev**
Default bind is ]. 

In the member list, yellow = leader, blue = self. Leader overrides self.

Bit of a preview:
Theoretically this should be significantly less laggy without 2 instances open while recording on a decently old laptop

https://user-images.githubusercontent.com/57310593/171276980-ccd87cc3-c104-454a-82e8-527a9c49362e.mp4

### CODE REVIEW

File: src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        Collections.sort(temporaryMembers);
+        partyMembers.clear();
+
+        // We put ourselves at the top, owner as #2
+        String playerName = McIf.player().getName();
+        String ownerName = PlayerInfo.get(SocialData.class).getPlayerParty().getOwner();
+        if (ownerName.equals("")) return; // We are not in a party
+
+        temporaryMembers.remove(playerName);
+        temporaryMembers.remove(ownerName);
```
Comment (Incompleteusern) [inline src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java]: check for changes before redoing all the work

File: src/main/java/com/wynntils/modules/core/managers/PartyManager.java [hunk 2]
```diff
@@ -42,36 +42,41 @@ public static void handlePartyList() {
...
             String member = McIf.getUnformattedText(component).split(" has left the party.")[0];
             partyContainer.removeMember(member);
+            return;
+        }
+        if (unformattedText.startsWith("You are now the leader of this party! Type /party for a list of commands.")) {
+            partyContainer.setOwner(McIf.player().getName());
+            return;
+        }
+        if (unformattedText.startsWith("Successfully promoted ") && unformattedText.endsWith(" to party leader!")) {
+            partyContainer.setOwner(unformattedText.split(" ")[2]);
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/core/managers/PartyManager.java]: Regex for this?

File: src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            } else if (playerName.equals(McIf.player().getName())) {
+                colour = 0x00FFFF;
+            } else {
+                colour = 0xFFFFFF;
+            }
+
+            fontRenderer.drawString(playerName, this.width/2 - 165, (verticalReference + 17) + 25 * i, colour);
+            // Blank text to reset colour to white
+            // This is so player heads don't have the previous self/owner colour overlayed onto them
+            fontRenderer.drawString("", 0, 0, 0xFFFFFF);
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java]: Can't you reset  the color with `GlStateManager#color(r,g,b,a)`

File: src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    private void invitePlayersFromTextField() {
+        // Remove anything that isn't possible in a minecraft name, except commas and spaces
+        String fieldText = inviteField.getText().replaceAll("[^\\w, ]+", "");
+        fieldText = fieldText.replaceAll("[, ]+", ","); // commas and spaces to just comma
+        if (fieldText.equals("")) return;
+
+        // Create party if we aren't in one
+        if (PlayerInfo.get(SocialData.class).getPlayerParty().getOwner().equals("")) {
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java]: Bit of a strange way to check if we are in a party? Maybe make a helper method in `PartyManager`?

File: src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+        temporaryMembers.remove(playerName);
+        temporaryMembers.remove(ownerName);
+
+        partyMembers.add(ownerName);
+        if (!ownerName.equals(playerName)) {
+            partyMembers.add(playerName);
+        }
+        partyMembers.addAll(temporaryMembers);
+        // List is now sorted in [self, owner, party alphabetically]
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java]: Perhaps a custom sorting lamba?

**Comment by DevScyu** [comment]
I noticed when watching the video, the party auto disbands when you kick the last remaining member other than yourself:
image

**Comment by ryanzhoudev** [comment]
> I noticed when watching the video, the party auto disbands when you kick the last remaining member other than yourself: image

The party auto disbands when the last person in the party (yourself) leaves. Wynn doesn't disband if eg. v8j kicks elianazhao and v8j is the last player remaining. If v8j then leaves, the party will disband. Also I believe that message is handled properly in PartyManager.java L54

File: src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java (hunk 4 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/utilities/overlays/ui/PartyManagementUI.java]: Oh apparently a method for this already exists, i'll switch to it

**Review by kristofbolyai — APPROVED** [review]

**Review by magicus — APPROVED** [review]

**Pull request merged**